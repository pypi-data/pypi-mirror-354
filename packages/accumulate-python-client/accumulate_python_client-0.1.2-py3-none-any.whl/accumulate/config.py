# accumulate/config.py

# Define Accumulate network base URLs (without "/v3")
# ACCUMULATE_RPC_URL = "https://testnet.accumulatenetwork.io"  # Default: Testnet

ACCUMULATE_RPC_URL = "https://mainnet.accumulatenetwork.io"  # Default: Testnet

# Optional: Add mainnet and custom options
ACCUMULATE_MAINNET_URL = "https://mainnet.accumulatenetwork.io"
ACCUMULATE_TESTNET_URL = "https://testnet.accumulatenetwork.io"
ACCUMULATE_CUSTOM_URL = None  # Set this if using a custom Accumulate node

# Function to get the active Accumulate RPC URL
def get_accumulate_rpc_url():
    """Returns the currently selected Accumulate RPC URL."""
    return ACCUMULATE_CUSTOM_URL or ACCUMULATE_RPC_URL
