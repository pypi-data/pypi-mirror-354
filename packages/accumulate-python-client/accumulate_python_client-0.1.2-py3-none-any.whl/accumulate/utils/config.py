# accumulate-python-client\accumulate\utils\config.py

import os

class Config:
    """
    Configuration utility for managing environment-specific settings.
    """
    @staticmethod
    def is_testnet() -> bool:
        """
        Dynamically get the testnet status based on the environment variable.
        """
        return os.getenv("ACCUMULATE_IS_TESTNET", "false").lower() == "true"

    @staticmethod
    def initial_acme_oracle() -> float:
        """
        Dynamically get the initial ACME oracle value based on the network type.
        """
        return 5000 if Config.is_testnet() else 0.50

    @staticmethod
    def get_network_type() -> str:
        """
        Get the current network type.
        :return: 'testnet' or 'mainnet'
        """
        return "testnet" if Config.is_testnet() else "mainnet"

    @staticmethod
    def get_initial_oracle_value() -> float:
        """
        Get the initial ACME oracle value for the configured network.
        :return: The oracle value as a float.
        """
        return Config.initial_acme_oracle()
