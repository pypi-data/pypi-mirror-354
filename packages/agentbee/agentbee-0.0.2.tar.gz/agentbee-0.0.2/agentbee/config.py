import configparser
from pathlib import Path

CONFIG_DIR = Path.home() / ".agentbee"
CONFIG_FILE = CONFIG_DIR / "config.ini"
CONFIG_SECTION = "DEFAULT"

def get_config_path() -> Path:
    """Returns the path to the config file."""
    return CONFIG_FILE

def save_config(api_key: str, base_url: str, model: str):
    """Saves the configuration to the global config file."""
    CONFIG_DIR.mkdir(exist_ok=True)
    config = configparser.ConfigParser()
    config[CONFIG_SECTION] = {
        'llm_api_key': api_key,
        'llm_base_url': base_url,
        'llm_model': model
    }
    with open(CONFIG_FILE, 'w') as configfile:
        config.write(configfile)
    print(f"âœ… Configuration saved to {CONFIG_FILE}")

def load_config() -> dict:
    if not CONFIG_FILE.exists():
        return {}
    config = configparser.ConfigParser()
    config.read(CONFIG_FILE)
    return {
        'llm_api_key': config.get(CONFIG_SECTION, 'llm_api_key', fallback=''),
        'llm_base_url': config.get(CONFIG_SECTION, 'llm_base_url', fallback=''),
        'llm_model': config.get(CONFIG_SECTION, 'llm_model', fallback='')
    }