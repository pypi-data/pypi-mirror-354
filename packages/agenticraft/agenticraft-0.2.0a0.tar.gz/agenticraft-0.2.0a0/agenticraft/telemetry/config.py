"""Telemetry configuration for AgentiCraft.

This module provides configuration management for OpenTelemetry
settings, including exporters, sampling, and instrumentation options.

Example:
    Basic configuration::

        from agenticraft.telemetry import TelemetryConfig

        config = TelemetryConfig(
            service_name="my-agent-service",
            export_endpoint="http://jaeger:4317",
            sample_rate=0.1,  # Sample 10% of traces
            export_metrics=True,
            export_traces=True
        )
"""

import os
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, model_validator


class ExportFormat(str, Enum):
    """Supported telemetry export formats."""

    OTLP = "otlp"
    JAEGER = "jaeger"
    ZIPKIN = "zipkin"
    CONSOLE = "console"
    NONE = "none"


class TelemetryEnvironment(str, Enum):
    """Deployment environment for telemetry."""

    DEVELOPMENT = "development"
    STAGING = "staging"
    PRODUCTION = "production"
    TEST = "test"


class ExporterConfig(BaseModel):
    """Configuration for telemetry exporters."""

    format: ExportFormat = Field(
        default=ExportFormat.OTLP, description="Export format to use"
    )
    endpoint: str | None = Field(default=None, description="Endpoint for exporter")
    headers: dict[str, str] = Field(
        default_factory=dict, description="Headers to include with exports"
    )
    timeout_ms: int = Field(default=30000, description="Export timeout in milliseconds")
    insecure: bool = Field(
        default=True, description="Whether to use insecure connection"
    )

    @model_validator(mode="after")
    def set_endpoint_defaults(self):
        """Set endpoint defaults based on format."""
        if self.endpoint is None and self.format != ExportFormat.CONSOLE:
            format_map = {
                ExportFormat.OTLP: "http://localhost:4317",
                ExportFormat.JAEGER: "http://localhost:14250",
                ExportFormat.ZIPKIN: "http://localhost:9411/api/v2/spans",
            }
            self.endpoint = format_map.get(self.format)
        return self


class ResourceConfig(BaseModel):
    """Resource attributes for telemetry."""

    service_name: str = Field(default="agenticraft", description="Name of the service")
    service_version: str = Field(
        default="unknown", description="Version of the service"
    )
    service_instance_id: str | None = Field(
        default=None, description="Unique instance identifier"
    )
    environment: TelemetryEnvironment = Field(
        default=TelemetryEnvironment.DEVELOPMENT, description="Deployment environment"
    )
    attributes: dict[str, Any] = Field(
        default_factory=dict, description="Additional resource attributes"
    )


class SamplingConfig(BaseModel):
    """Sampling configuration for traces."""

    sample_rate: float = Field(
        default=1.0, ge=0.0, le=1.0, description="Base sampling rate (0.0-1.0)"
    )
    parent_based: bool = Field(
        default=True, description="Honor parent span sampling decision"
    )
    rules: list[dict[str, Any]] = Field(
        default_factory=list, description="Custom sampling rules"
    )


class InstrumentationConfig(BaseModel):
    """Configuration for automatic instrumentation."""

    instrument_http: bool = Field(default=True, description="Instrument HTTP requests")
    instrument_grpc: bool = Field(default=True, description="Instrument gRPC calls")
    instrument_redis: bool = Field(
        default=False, description="Instrument Redis operations"
    )
    instrument_sqlalchemy: bool = Field(
        default=False, description="Instrument database queries"
    )
    excluded_urls: list[str] = Field(
        default_factory=lambda: ["/health", "/metrics"],
        description="URLs to exclude from tracing",
    )


class TelemetryConfig(BaseModel):
    """Complete telemetry configuration."""

    enabled: bool = Field(default=True, description="Whether telemetry is enabled")

    # Resource configuration
    resource: ResourceConfig = Field(
        default_factory=ResourceConfig, description="Resource configuration"
    )

    # Tracing configuration
    export_traces: bool = Field(default=True, description="Whether to export traces")
    trace_exporter: ExporterConfig = Field(
        default_factory=ExporterConfig, description="Trace exporter configuration"
    )
    sampling: SamplingConfig = Field(
        default_factory=SamplingConfig, description="Trace sampling configuration"
    )

    # Metrics configuration
    export_metrics: bool = Field(default=True, description="Whether to export metrics")
    metric_exporter: ExporterConfig = Field(
        default_factory=ExporterConfig, description="Metric exporter configuration"
    )
    metric_interval_ms: int = Field(
        default=60000, description="Metric export interval in milliseconds"
    )

    # Instrumentation
    instrumentation: InstrumentationConfig = Field(
        default_factory=InstrumentationConfig,
        description="Automatic instrumentation settings",
    )

    # Logging integration
    log_correlation: bool = Field(default=True, description="Add trace IDs to logs")

    @classmethod
    def from_env(cls) -> "TelemetryConfig":
        """Create configuration from environment variables.

        Supports standard OpenTelemetry environment variables:
        - OTEL_SERVICE_NAME
        - OTEL_EXPORTER_OTLP_ENDPOINT
        - OTEL_TRACES_EXPORTER
        - OTEL_METRICS_EXPORTER
        - OTEL_TRACES_SAMPLER_ARG

        Returns:
            Configured TelemetryConfig instance
        """
        config = cls()

        # Service configuration
        if service_name := os.getenv("OTEL_SERVICE_NAME"):
            config.resource.service_name = service_name

        if service_version := os.getenv("OTEL_SERVICE_VERSION"):
            config.resource.service_version = service_version

        # Endpoint configuration
        if endpoint := os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT"):
            config.trace_exporter.endpoint = endpoint
            config.metric_exporter.endpoint = endpoint

        # Trace-specific endpoint
        if trace_endpoint := os.getenv("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT"):
            config.trace_exporter.endpoint = trace_endpoint

        # Metric-specific endpoint
        if metric_endpoint := os.getenv("OTEL_EXPORTER_OTLP_METRICS_ENDPOINT"):
            config.metric_exporter.endpoint = metric_endpoint

        # Headers
        if headers := os.getenv("OTEL_EXPORTER_OTLP_HEADERS"):
            parsed_headers = {}
            for header in headers.split(","):
                if "=" in header:
                    key, value = header.split("=", 1)
                    parsed_headers[key.strip()] = value.strip()
            config.trace_exporter.headers = parsed_headers
            config.metric_exporter.headers = parsed_headers

        # Sampling
        if sample_rate := os.getenv("OTEL_TRACES_SAMPLER_ARG"):
            try:
                config.sampling.sample_rate = float(sample_rate)
            except ValueError:
                pass

        # Exporters
        if traces_exporter := os.getenv("OTEL_TRACES_EXPORTER"):
            if traces_exporter == "none":
                config.export_traces = False
            elif traces_exporter in ExportFormat.__members__.values():
                config.trace_exporter.format = ExportFormat(traces_exporter)

        if metrics_exporter := os.getenv("OTEL_METRICS_EXPORTER"):
            if metrics_exporter == "none":
                config.export_metrics = False
            elif metrics_exporter in ExportFormat.__members__.values():
                config.metric_exporter.format = ExportFormat(metrics_exporter)

        return config

    def to_env_dict(self) -> dict[str, str]:
        """Convert configuration to environment variables.

        Returns:
            Dictionary of environment variables
        """
        env_vars = {
            "OTEL_SERVICE_NAME": self.resource.service_name,
            "OTEL_SERVICE_VERSION": self.resource.service_version,
            "OTEL_TRACES_SAMPLER_ARG": str(self.sampling.sample_rate),
        }

        if self.trace_exporter.endpoint:
            env_vars["OTEL_EXPORTER_OTLP_ENDPOINT"] = self.trace_exporter.endpoint

        if self.trace_exporter.headers:
            headers_str = ",".join(
                f"{k}={v}" for k, v in self.trace_exporter.headers.items()
            )
            env_vars["OTEL_EXPORTER_OTLP_HEADERS"] = headers_str

        if not self.export_traces:
            env_vars["OTEL_TRACES_EXPORTER"] = "none"
        else:
            env_vars["OTEL_TRACES_EXPORTER"] = self.trace_exporter.format.value

        if not self.export_metrics:
            env_vars["OTEL_METRICS_EXPORTER"] = "none"
        else:
            env_vars["OTEL_METRICS_EXPORTER"] = self.metric_exporter.format.value

        return env_vars


# Preset configurations for common scenarios


def development_config() -> TelemetryConfig:
    """Get development environment configuration.

    - Console exporter for easy debugging
    - Full sampling
    - All instrumentation enabled
    """
    return TelemetryConfig(
        resource=ResourceConfig(environment=TelemetryEnvironment.DEVELOPMENT),
        trace_exporter=ExporterConfig(format=ExportFormat.CONSOLE),
        metric_exporter=ExporterConfig(format=ExportFormat.CONSOLE),
        sampling=SamplingConfig(sample_rate=1.0),
    )


def production_config(
    service_name: str, otlp_endpoint: str, sample_rate: float = 0.1
) -> TelemetryConfig:
    """Get production environment configuration.

    Args:
        service_name: Name of your service
        otlp_endpoint: OTLP collector endpoint
        sample_rate: Sampling rate (default 10%)

    Returns:
        Production-ready configuration
    """
    return TelemetryConfig(
        resource=ResourceConfig(
            service_name=service_name, environment=TelemetryEnvironment.PRODUCTION
        ),
        trace_exporter=ExporterConfig(
            format=ExportFormat.OTLP, endpoint=otlp_endpoint, insecure=False
        ),
        metric_exporter=ExporterConfig(
            format=ExportFormat.OTLP, endpoint=otlp_endpoint, insecure=False
        ),
        sampling=SamplingConfig(sample_rate=sample_rate),
    )


def test_config() -> TelemetryConfig:
    """Get test environment configuration.

    - Telemetry disabled by default
    - No external dependencies
    """
    return TelemetryConfig(
        enabled=False, resource=ResourceConfig(environment=TelemetryEnvironment.TEST)
    )
