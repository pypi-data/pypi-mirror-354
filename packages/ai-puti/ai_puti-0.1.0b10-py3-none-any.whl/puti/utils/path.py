"""
@Author: obstacle
@Time: 10/01/25 14:03
@Description:  
"""
from pathlib import Path
from loguru import logger


def root_dir():
    package_root = Path(__file__).parent.parent.parent
    for i in (".git", ".project_root", ".gitignore"):
        if (package_root / i).exists():
            break
        else:
            package_root = Path.cwd()
    # logger.info(f'Package root set to {str(package_root)}')
    return package_root
