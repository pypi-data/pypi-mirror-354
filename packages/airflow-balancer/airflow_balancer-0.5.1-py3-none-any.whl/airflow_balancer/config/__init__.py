from .balancer import *
from .host import *
from .port import *
