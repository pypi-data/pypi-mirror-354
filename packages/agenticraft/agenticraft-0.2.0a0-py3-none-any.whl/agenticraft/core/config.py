"""Configuration management for AgentiCraft.

This module provides centralized configuration management using
pydantic-settings. Configuration can be loaded from environment
variables, .env files, or set programmatically.

Example:
    Loading configuration::

        from agenticraft.core.config import settings

        # Access configuration
        print(settings.openai_api_key)
        print(settings.default_model)

        # Override configuration
        settings.default_temperature = 0.5
"""

from functools import lru_cache
from pathlib import Path
from typing import Any

from pydantic import Field, field_validator
from pydantic_settings import BaseSettings, SettingsConfigDict


class AgentiCraftSettings(BaseSettings):
    """Global settings for AgentiCraft.

    Settings are loaded from (in order of priority):
    1. Environment variables
    2. .env file in current directory
    3. .env file in parent directories
    4. Default values

    All settings can be overridden using environment variables
    with the AGENTICRAFT_ prefix.
    """

    model_config = SettingsConfigDict(
        env_prefix="AGENTICRAFT_",
        env_file=".env",
        env_file_encoding="utf-8",
        case_sensitive=False,
        extra="allow",
    )

    # General settings
    environment: str = Field(
        default="development",
        description="Environment (development, staging, production)",
    )
    debug: bool = Field(default=False, description="Enable debug mode")
    log_level: str = Field(default="INFO", description="Logging level")

    # API Keys
    openai_api_key: str | None = Field(default=None, description="OpenAI API key")
    anthropic_api_key: str | None = Field(default=None, description="Anthropic API key")
    google_api_key: str | None = Field(default=None, description="Google API key")

    # Model defaults
    default_model: str = Field(default="gpt-4", description="Default LLM model")
    default_temperature: float = Field(
        default=0.7, ge=0.0, le=2.0, description="Default temperature for LLM"
    )
    default_max_tokens: int | None = Field(
        default=None, gt=0, description="Default max tokens for LLM"
    )
    default_timeout: int = Field(
        default=30, gt=0, description="Default timeout in seconds"
    )
    default_max_retries: int = Field(default=3, ge=0, description="Default max retries")

    # Provider settings
    openai_base_url: str = Field(
        default="https://api.openai.com/v1", description="OpenAI API base URL"
    )
    anthropic_base_url: str = Field(
        default="https://api.anthropic.com", description="Anthropic API base URL"
    )
    ollama_base_url: str = Field(
        default="http://localhost:11434", description="Ollama base URL"
    )

    # Memory settings
    memory_backend: str = Field(
        default="sqlite", description="Memory backend (sqlite, json, redis)"
    )
    memory_path: Path = Field(
        default=Path("./memory"), description="Path for memory storage"
    )
    conversation_memory_size: int = Field(
        default=10, gt=0, description="Number of conversation turns to keep"
    )

    # Telemetry settings
    telemetry_enabled: bool = Field(default=True, description="Enable telemetry")
    telemetry_service_name: str = Field(
        default="agenticraft", description="Service name for telemetry"
    )
    telemetry_export_endpoint: str | None = Field(
        default=None, description="OTLP endpoint for telemetry export"
    )
    telemetry_sample_rate: float = Field(
        default=1.0, ge=0.0, le=1.0, description="Telemetry sampling rate"
    )

    # Plugin settings
    plugins_enabled: bool = Field(default=True, description="Enable plugin system")
    plugin_dirs: list[Path] = Field(
        default_factory=lambda: [Path("./plugins")],
        description="Directories to load plugins from",
    )

    # MCP settings
    mcp_enabled: bool = Field(default=True, description="Enable Model Context Protocol")
    mcp_servers: list[str] = Field(
        default_factory=list, description="MCP server URLs to connect to"
    )

    # Tool settings
    tool_execution_timeout: int = Field(
        default=10, gt=0, description="Timeout for tool execution in seconds"
    )
    tool_retry_attempts: int = Field(
        default=2, ge=0, description="Number of retry attempts for failed tools"
    )

    # Workflow settings
    workflow_step_timeout: int = Field(
        default=60, gt=0, description="Default timeout for workflow steps in seconds"
    )
    workflow_max_parallel_steps: int = Field(
        default=5, gt=0, description="Maximum parallel steps in workflows"
    )

    # Security settings
    allow_code_execution: bool = Field(
        default=False, description="Allow execution of arbitrary code in tools"
    )
    allowed_domains: set[str] = Field(
        default_factory=set, description="Allowed domains for web requests"
    )

    @field_validator("log_level")
    def validate_log_level(cls, v: str) -> str:
        """Validate log level."""
        valid_levels = {"DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"}
        v_upper = v.upper()
        if v_upper not in valid_levels:
            raise ValueError(f"Invalid log level: {v}. Must be one of {valid_levels}")
        return v_upper

    @field_validator("environment")
    def validate_environment(cls, v: str) -> str:
        """Validate environment."""
        valid_envs = {"development", "staging", "production", "test"}
        v_lower = v.lower()
        if v_lower not in valid_envs:
            raise ValueError(f"Invalid environment: {v}. Must be one of {valid_envs}")
        return v_lower

    @field_validator("memory_backend")
    def validate_memory_backend(cls, v: str) -> str:
        """Validate memory backend."""
        valid_backends = {"sqlite", "json", "redis"}
        v_lower = v.lower()
        if v_lower not in valid_backends:
            raise ValueError(
                f"Invalid memory backend: {v}. Must be one of {valid_backends}"
            )
        return v_lower

    def get_api_key(self, provider: str) -> str | None:
        """Get API key for a provider.

        Args:
            provider: Provider name (openai, anthropic, google)

        Returns:
            API key if available
        """
        key_map = {
            "openai": self.openai_api_key,
            "anthropic": self.anthropic_api_key,
            "google": self.google_api_key,
        }
        return key_map.get(provider.lower())

    def get_base_url(self, provider: str) -> str | None:
        """Get base URL for a provider.

        Args:
            provider: Provider name

        Returns:
            Base URL if available
        """
        url_map = {
            "openai": self.openai_base_url,
            "anthropic": self.anthropic_base_url,
            "ollama": self.ollama_base_url,
        }
        return url_map.get(provider.lower())

    def to_agent_config(self) -> dict[str, Any]:
        """Convert settings to agent configuration.

        Returns:
            Dictionary suitable for Agent initialization
        """
        return {
            "model": self.default_model,
            "temperature": self.default_temperature,
            "max_tokens": self.default_max_tokens,
            "timeout": self.default_timeout,
            "max_retries": self.default_max_retries,
        }

    def to_telemetry_config(self) -> dict[str, Any]:
        """Convert settings to telemetry configuration.

        Returns:
            Dictionary suitable for Telemetry initialization
        """
        return {
            "service_name": self.telemetry_service_name,
            "export_to": self.telemetry_export_endpoint,
            "enabled": self.telemetry_enabled,
            "sample_rate": self.telemetry_sample_rate,
        }

    @property
    def is_production(self) -> bool:
        """Check if running in production."""
        return self.environment == "production"

    @property
    def is_development(self) -> bool:
        """Check if running in development."""
        return self.environment == "development"

    def validate_required_keys(self, providers: list[str]) -> None:
        """Validate that required API keys are present.

        Args:
            providers: List of providers that need keys

        Raises:
            ValueError: If required keys are missing
        """
        missing = []
        for provider in providers:
            if not self.get_api_key(provider):
                missing.append(provider)

        if missing:
            raise ValueError(
                f"Missing API keys for providers: {', '.join(missing)}. "
                f"Set environment variables or update .env file."
            )


@lru_cache
def get_settings() -> AgentiCraftSettings:
    """Get the global settings instance (cached).

    Returns:
        AgentiCraftSettings instance
    """
    return AgentiCraftSettings()


# Global settings instance
settings = get_settings()


def reload_settings() -> AgentiCraftSettings:
    """Reload settings from environment.

    This clears the cache and creates a new settings instance.
    Useful for testing or when environment variables change.

    Returns:
        New AgentiCraftSettings instance
    """
    get_settings.cache_clear()
    global settings
    settings = get_settings()
    return settings


def update_settings(**kwargs: Any) -> None:
    """Update settings programmatically.

    Args:
        **kwargs: Setting values to update

    Example:
        update_settings(default_model="gpt-3.5-turbo", debug=True)
    """
    for key, value in kwargs.items():
        if hasattr(settings, key):
            setattr(settings, key, value)
        else:
            raise ValueError(f"Unknown setting: {key}")
