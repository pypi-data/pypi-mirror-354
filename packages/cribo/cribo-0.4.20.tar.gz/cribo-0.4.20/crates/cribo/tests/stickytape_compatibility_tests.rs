#![allow(clippy::disallowed_methods)]

/// Comprehensive test suite based on stickytape test scenarios
///
/// This test suite mirrors the functionality and test cases from the stickytape Python bundler
/// to ensure our Rust implementation (serpen) can handle all the same scenarios correctly.
///
/// Tests are organized to match the original stickytape test structure:
/// - Single file scripts (no bundling needed)
/// - Scripts with stdlib imports (should not bundle stdlib modules)
/// - Scripts with local imports (should bundle local modules)
/// - Scripts with various import syntaxes
/// - Scripts with relative imports
/// - Scripts with special edge cases
///
/// Reference: https://github.com/mwilliamson/stickytape
use cow_utils::CowUtils;
use insta::{assert_snapshot, with_settings};
use std::path::PathBuf;
use std::process::Command;
use tempfile::TempDir;

use cribo::bundler::Bundler;
use cribo::config::Config;

/// Base path to the copied stickytape test scripts
const STICKYTAPE_FIXTURES: &str = "tests/fixtures/stickytape_test_scripts";

/// Helper function to run a bundled script and capture its output
fn run_bundled_script(script_path: &PathBuf) -> Result<String, Box<dyn std::error::Error>> {
    let output = Command::new("python3").arg(script_path).output()?;

    // Debug: Check exit status and stderr
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        eprintln!("Script execution failed with status: {}", output.status);
        eprintln!("Stderr: {}", stderr);
        return Err(format!("Script execution failed: {}", stderr).into());
    }

    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    let stderr = String::from_utf8_lossy(&output.stderr);

    if !stderr.is_empty() {
        eprintln!("Script stderr: {}", stderr);
    }

    Ok(stdout)
}

/// Helper function to bundle a test script and return the bundled content and output path
fn bundle_test_script(
    script_name: &str,
) -> Result<(String, PathBuf, TempDir), Box<dyn std::error::Error>> {
    let script_dir = PathBuf::from(STICKYTAPE_FIXTURES).join(script_name);
    let entry_script = script_dir.join("hello");

    // Skip if the test fixture doesn't exist
    if !entry_script.exists() {
        return Err(format!("Test script not found: {:?}", entry_script).into());
    }

    let config = Config {
        src: vec![script_dir],
        ..Default::default()
    };
    let mut bundler = Bundler::new(config);

    let temp_dir = TempDir::new()?;
    let output_path = temp_dir.path().join("bundled_script.py");

    // Bundle the script
    bundler.bundle(&entry_script, &output_path, false)?;

    // Read the bundled content
    let bundled_content = std::fs::read_to_string(&output_path)?;

    // Return temp_dir to keep it alive
    Ok((bundled_content, output_path, temp_dir))
}

/// Helper function to test script output matches expected result
fn assert_script_output(script_name: &str, expected_output: &str) {
    let result = bundle_test_script(script_name);

    match result {
        Ok((bundled_content, output_path, _temp_dir)) => {
            // Debug: Print bundled content to understand what's happening
            eprintln!("=== BUNDLED CONTENT FOR {} ===", script_name);
            eprintln!("{}", bundled_content);
            eprintln!("=== END BUNDLED CONTENT ===");

            // Verify the script runs and produces expected output
            match run_bundled_script(&output_path) {
                Ok(actual_output) => {
                    eprintln!("Script output: '{}'", actual_output);
                    eprintln!("Expected output: '{}'", expected_output);

                    // Normalize both outputs by trimming whitespace and normalizing line endings
                    let actual_normalized = actual_output.trim().cow_replace("\r\n", "\n");
                    let expected_normalized = expected_output.trim().cow_replace("\r\n", "\n");

                    assert_eq!(
                        actual_normalized, expected_normalized,
                        "Script output mismatch for {}",
                        script_name
                    );

                    // Create snapshot for the bundled content
                    with_settings!({
                        description => format!("Bundled content for {}", script_name),
                        omit_expression => true,
                    }, {
                        assert_snapshot!(format!("bundled_{}", script_name.cow_replace("/", "_")), bundled_content);
                    });
                }
                Err(e) => {
                    panic!(
                        "Failed to execute bundled script for {}: {}",
                        script_name, e
                    );
                }
            }
        }
        Err(e) => {
            // If the test fixture doesn't exist, skip the test
            if e.to_string().contains("Test script not found") {
                eprintln!("Skipping test for {}: {}", script_name, e);
                return;
            }
            panic!("Failed to bundle script {}: {}", script_name, e);
        }
    }
}

#[test]
fn test_single_file_script_still_works() {
    assert_script_output("single_file", "Hello");
}

#[test]
fn test_stdlib_imports_are_not_modified() {
    assert_script_output(
        "single_file_using_stdlib",
        "f7ff9e8b7bb2e09b70935a5d785e0cc5d9d0abf0",
    );
}

#[test]
fn test_stdlib_module_in_package_is_not_generated() {
    assert_script_output(
        "script_using_stdlib_module_in_package",
        "xml.etree.ElementTree\nHello",
    );
}

#[test]
fn test_script_that_imports_local_module_is_converted_to_single_file() {
    assert_script_output("script_with_single_local_import", "Hello");
}

#[test]
fn test_script_that_imports_local_package_is_converted_to_single_file() {
    assert_script_output("script_with_single_local_import_of_package", "Hello");
}

#[test]
fn test_can_import_module_from_package() {
    assert_script_output("script_using_module_in_package", "Hello");
}

#[test]
fn test_can_import_value_from_module_using_from_import_syntax() {
    assert_script_output("script_with_single_local_from_import", "Hello");
}

#[test]
fn test_can_import_multiple_values_from_module_using_from_import_syntax() {
    assert_script_output("script_using_from_to_import_multiple_values", "Hello");
}

#[test]
fn test_can_import_module_from_package_using_from_import_syntax() {
    assert_script_output("script_using_from_to_import_module", "Hello");
}

#[test]
fn test_imported_modules_are_transformed() {
    assert_script_output("imports_in_imported_modules", "Hello");
}

#[test]
fn test_circular_references_dont_cause_stack_overflow() {
    assert_script_output("circular_reference", "Hello");
}

#[test]
fn test_explicit_relative_imports_with_single_dot_are_resolved_correctly() {
    assert_script_output("explicit_relative_import_single_dot", "Hello");
}

#[test]
fn test_explicit_relative_imports_with_single_dot_in_package_init_are_resolved_correctly() {
    assert_script_output("explicit_relative_import_single_dot_in_init", "Hello");
}

#[test]
fn test_explicit_relative_imports_from_parent_package_are_resolved_correctly() {
    assert_script_output("explicit_relative_import_from_parent_package", "Hello");
}

#[test]
fn test_explicit_relative_imports_with_module_name_are_resolved_correctly() {
    assert_script_output("explicit_relative_import", "Hello");
}

#[test]
fn test_explicit_relative_imports_with_module_name_in_package_init_are_resolved_correctly() {
    assert_script_output("explicit_relative_import_in_init", "Hello");
}

#[test]
fn test_package_init_can_be_used_even_if_not_imported_explicitly() {
    assert_script_output("implicit_init_import", "Hello");
}

#[test]
fn test_value_import_is_detected_when_import_is_renamed() {
    assert_script_output("import_from_as_value", "Hello");
}

#[test]
fn test_module_import_is_detected_when_import_is_renamed() {
    assert_script_output("import_from_as_module", "Hello");
}

#[test]
fn test_modules_with_triple_quotes_can_be_bundled() {
    assert_script_output("module_with_triple_quotes", "Hello\n'''\n\"\"\"");
}

#[test]
fn test_additional_python_modules_can_be_explicitly_included() {
    // This test is for dynamic imports which may require special handling
    // We'll test if we can bundle scripts with dynamic imports
    let script_name = "script_with_dynamic_import";
    let result = bundle_test_script(script_name);

    match result {
        Ok((bundled_content, _output_path, _temp_dir)) => {
            // Just verify that bundling succeeds for now
            // Dynamic import handling may need special configuration
            with_settings!({
                description => format!("Bundled content for {}", script_name),
                omit_expression => true,
            }, {
                assert_snapshot!("bundled_script_with_dynamic_import", bundled_content);
            });
        }
        Err(e) => {
            if e.to_string().contains("Test script not found") {
                eprintln!("Skipping test for {}: {}", script_name, e);
                return;
            }
            // For now, we'll allow this to fail since dynamic imports are complex
            eprintln!("Dynamic import test failed (expected): {}", e);
        }
    }
}

/// Test that our bundler can handle special shebang preservation
#[test]
fn test_special_shebang_handling() {
    let script_name = "script_with_special_shebang";
    let result = bundle_test_script(script_name);

    match result {
        Ok((bundled_content, _output_path, _temp_dir)) => {
            // Verify the bundled content preserves or handles shebangs appropriately
            with_settings!({
                description => format!("Bundled content with shebang for {}", script_name),
                omit_expression => true,
            }, {
                assert_snapshot!("bundled_script_with_special_shebang", bundled_content);
            });
        }
        Err(e) => {
            if e.to_string().contains("Test script not found") {
                eprintln!("Skipping test for {}: {}", script_name, e);
                return;
            }
            panic!("Failed to bundle script with special shebang: {}", e);
        }
    }
}

/// Integration test to verify our bundler handles all major scenarios
#[test]
fn test_comprehensive_bundling_scenarios() {
    let _ = env_logger::try_init();

    let test_scenarios = vec![
        ("single_file", "Single file without dependencies"),
        (
            "script_with_single_local_import",
            "Script with local module import",
        ),
        (
            "script_using_module_in_package",
            "Script importing from package",
        ),
        (
            "script_with_single_local_from_import",
            "Script with from-import syntax",
        ),
        ("imports_in_imported_modules", "Nested module imports"),
    ];

    let mut results = Vec::new();
    let mut successful_bundles = 0;
    let mut failed_bundles = 0;

    for (scenario, description) in &test_scenarios {
        match bundle_test_script(scenario) {
            Ok((bundled_content, output_path, _temp_dir)) => {
                let lines = bundled_content.lines().count();
                let size = bundled_content.len();

                // Try to run the script
                let execution_result = run_bundled_script(&output_path);
                let execution_status = match execution_result {
                    Ok(output) => format!("Success: {}", output.trim()),
                    Err(e) => format!("Failed: {}", e),
                };

                results.push(format!(
                    "✓ {} ({})\n  Lines: {}, Size: {} bytes\n  Execution: {}",
                    scenario, description, lines, size, execution_status
                ));
                successful_bundles += 1;
            }
            Err(e) => {
                if e.to_string().contains("Test script not found") {
                    results.push(format!(
                        "⏭ {} ({}): Skipped - fixture not found",
                        scenario, description
                    ));
                } else {
                    results.push(format!("✗ {} ({}): {}", scenario, description, e));
                    failed_bundles += 1;
                }
            }
        }
    }

    let summary = format!(
        "Bundling Test Results Summary:\n{}\n\nSuccessful: {}, Failed: {}, Total scenarios: {}",
        results.join("\n\n"),
        successful_bundles,
        failed_bundles,
        test_scenarios.len()
    );

    with_settings!({
        description => "Comprehensive bundling test results across multiple scenarios",
    }, {
        assert_snapshot!("comprehensive_bundling_results", summary);
    });

    // The test should pass as long as we have some successful bundles
    assert!(
        successful_bundles > 0,
        "No scenarios were successfully bundled"
    );
}
