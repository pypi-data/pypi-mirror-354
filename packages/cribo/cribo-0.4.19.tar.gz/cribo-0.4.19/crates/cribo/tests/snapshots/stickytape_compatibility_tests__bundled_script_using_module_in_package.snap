---
source: crates/cribo/tests/stickytape_compatibility_tests.rs
description: Bundled content for script_using_module_in_package
---
#!/usr/bin/env python3
# Generated by Cribo - Python Source Bundler
# https://github.com/ophidiarium/cribo

# ─ Module: greetings.greeting ─
import types
greetings = types.ModuleType('greetings')
greetings.greeting = types.ModuleType('greetings.greeting')
exec('message = "Hello"', greetings.greeting.__dict__)

# ─ Entry Module: hello ─
print(greetings.greeting.message)
